libs/elasticsearch-nio/src/main/java/org/elasticsearch/nio/NioSelector.java

```diff
@@ -33,36 +34,68 @@
 import java.util.concurrent.Future;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiConsumer;
 import java.util.stream.Collectors;
 
 /**
- * This is a basic selector abstraction. This selector wraps a raw nio {@link Selector}. When you call
+ * This is a nio selector implementation. This selector wraps a raw nio {@link Selector}. When you call
  * {@link #runLoop()}, the selector will run until {@link #close()} is called. This instance handles closing
  * of channels. Users should call {@link #queueChannelClose(NioChannel)} to schedule a channel for close by
  * this selector.
  * <p>
  * Children of this class should implement the specific {@link #processKey(SelectionKey)},
  * {@link #preSelect()}, and {@link #cleanup()} functionality.
  */
-public abstract class ESSelector implements Closeable {
-
-    final Selector selector;
-    final ConcurrentLinkedQueue<ChannelContext<?>> channelsToClose = new ConcurrentLinkedQueue<>();
+public class NioSelector implements Closeable {
 
+    private final ConcurrentLinkedQueue<WriteOperation> queuedWrites = new ConcurrentLinkedQueue<>();
+    private final ConcurrentLinkedQueue<ChannelContext<?>> channelsToClose = new ConcurrentLinkedQueue<>();
+    private final ConcurrentLinkedQueue<ChannelContext<?>> channelsToRegister = new ConcurrentLinkedQueue<>();
     private final EventHandler eventHandler;
+    private final Selector selector;
+
     private final ReentrantLock runLock = new ReentrantLock();
     private final CountDownLatch exitedLoop = new CountDownLatch(1);
     private final AtomicBoolean isClosed = new AtomicBoolean(false);
     private final CompletableFuture<Void> isRunningFuture = new CompletableFuture<>();
     private volatile Thread thread;
 
-    ESSelector(EventHandler eventHandler) throws IOException {
+    public NioSelector(EventHandler eventHandler) throws IOException {
         this(eventHandler, Selector.open());
     }
 
-    ESSelector(EventHandler eventHandler, Selector selector) {
-        this.eventHandler = eventHandler;
+    public NioSelector(EventHandler eventHandler, Selector selector) throws IOException {
         this.selector = selector;
+        this.eventHandler = eventHandler;
+    }
+
+    public Selector rawSelector() {
+        return selector;
+    }
+
+    public boolean isOpen() {
+        return isClosed.get() == false;
+    }
+
+    public boolean isRunning() {
+        return runLock.isLocked();
+    }
+
+    Future<Void> isRunningFuture() {
+        return isRunningFuture;
+    }
+
+    void setThread() {
```

Comment:
can this be called by multiple threads? and / or should be make sure it's only set once?

Reply:
This is only default for tests. However, I added logic to make sure it is only set once.